File: backend/content-service/src/main/java/com/communityplatform/content/service/CommentService.java
package com.communityplatform.content.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.communityplatform.content.dto.comment.CommentCreateDto;
import com.communityplatform.content.dto.comment.CommentResponseDto;
import com.communityplatform.content.dto.comment.CommentUpdateDto;

import java.util.List;

/**
 * Service interface for Comment operations.
 */
public interface CommentService {

    /**
     * Create a new comment.
     */
    CommentResponseDto createComment(CommentCreateDto dto);

    /**
     * Get comment by ID.
     */
    CommentResponseDto getCommentById(Long commentId);

    /**
     * Update an existing comment.
     */
    CommentResponseDto updateComment(Long commentId, CommentUpdateDto dto, Long currentUserId);

    /**
     * Delete comment (soft delete).
     */
    void deleteComment(Long commentId, Long currentUserId);

    /**
     * Get all comments for a post.
     */
    Page<CommentResponseDto> getCommentsByPostId(Long postId, Pageable pageable);

    /**
     * Get top-level comments for a post (no parent).
     */
    Page<CommentResponseDto> getTopLevelComments(Long postId, Pageable pageable);

    /**
     * Get replies for a comment.
     */
    List<CommentResponseDto> getReplies(Long parentCommentId);

    /**
     * Increment like count.
     */
    void incrementLikeCount(Long commentId);

    /**
     * Decrement like count.
     */
    void decrementLikeCount(Long commentId);
}

File: frontend/src/app/core/services/comment.service.ts
import { inject, Injectable } from '@angular/core';
import { ApiService } from './api.service';
import { Comment, CommentCreate, CommentUpdate } from '../models/comment.model';
import { Page } from '../models/pagination.model';

@Injectable({ providedIn: 'root' })
export class CommentService {
  private api = inject(ApiService);

  create(postId: number, payload: CommentCreate) {
    return this.api.post<Comment>(`/v1/posts/${postId}/comments`, payload);
  }

  getById(commentId: number) {
    return this.api.get<Comment>(`/v1/comments/${commentId}`);
  }

  getByPost(postId: number, page = 0, size = 20) {
    return this.api.get<Page<Comment>>(`/v1/posts/${postId}/comments`, { page, size });
  }

  getTopLevel(postId: number, page = 0, size = 20) {
    return this.api.get<Page<Comment>>(`/v1/posts/${postId}/comments/top`, { page, size });
  }

  getReplies(commentId: number) {
    return this.api.get<Comment[]>(`/v1/comments/${commentId}/replies`);
  }

  update(commentId: number, payload: CommentUpdate) {
    return this.api.put<Comment>(`/v1/comments/${commentId}`, payload);
  }

  delete(commentId: number) {
    return this.api.delete<void>(`/v1/comments/${commentId}`);
  }
}
